File: Python/backjoon/9663.py
N = int(input())

board = [[0 for _ in range(N)] for _ in range(N)]

total = 0

directions = [
    (1, -1), (1, 1) # 윗쪽으로 가는건 없어도 무방
]

def add(y: int, x: int, flag: int) -> None:
    for i in range(N):
        board[i][x] += flag
        board[y][i] += flag

        for direction in directions:
            newY, newX = y + i * direction[0], x + i * direction[1]
            if 0 <= newY < N and 0 <= newX < N:
                board[newY][newX] += flag

def solve(y, x, count):
    global total

    if count == 0:
        total += 1
        return

    for j in range(0, N):
        if board[y][j] == 0:
            add(y, j, 1)
            solve(y + 1, j, count - 1)
            add(y, j, -1)

solve(-1, -1, N)
print(total)

File: CPP/Backjoon/9663.cpp
#include <iostream>
#include <cstring>
#include <string>
#include <vector>

std::vector< std::vector<int> > board;
int N;

int solve(std::vector<int>& participance, int count) {
    if (count == N)
        return 1;

    int res = 0;

    for (int i = 0; i < N; i++) {
        bool flag = true;
        for (int j = 0; j < participance.size(); j++) {
            if (i == participance[j] || std::abs(participance[j] - i) == (count - j)) {
                flag = false;
                break;
            }
        }

        if (flag) {
            participance.push_back(i);
            res += solve(participance, count + 1);
            participance.pop_back();
        }
    }

    return res;
}

int main() {
    std::cin >> N;

    for (int i = 0; i < N; i++)
        board.push_back(std::vector<int>(N, 0));

    std::vector<int> v;

    int result = solve(v, 0);

    std::cout << result << "\n";

    return 0;
}



/*
a: 1 2 4
b: 3 4 7

답: 1 2 3 4 7
*/